[DOCS] Add NatSpec documentation (#6122)
ethereum/solidity: docs/natspec-format.rst

##############
NatSpec Format
##############

Solidity contracts can use a special form of comments to provide rich
documentation for functions, return variables and more. This special form is
named the Ethereum Natural Language Specification Format (NatSpec).

> This documentation is segmented into developer-focused messages and end-user-

Reviewer: Will the newline create a space in the rendered document?

Author: This is currently undefined behavior as far as how I have written the documentation. But the actual implementation is Solidity is that newlines in the docblock are concatenated with the newline deleted. A full demonstration follows.

Our options at this point are:

1. Leave the issue of whitespace as undefined behavior for #6122 <kbd>RECOMMENDED</kbd>
2. Add whitespace handling to the specification

Please advise.

# Test case

```solidity
pragma solidity ^0.5.6;

/// @title A simulator for trees
/// extended title
/// @author Larry A. Gardner
///  extended author
/// @notice You can use this contract for only the most basic simulation
///         extended notice
/// @dev All function calls are currently implemented without side effects
/// @dev Repeated dev
contract Tree {
    /// @author Mary A. Botanist
    /// @notice Calculate tree age in years, rounded up, for live trees
    /// @dev The Alexandr N. Tetearing algorithm could increase precision
    /// @param rings The number of rings from dendrochronological sample
    /// @return age in years, rounded up for partial years
    function age(uint256 rings) external pure returns (uint256) {
        return rings + 1;
    }
}
```

Rendered output:

```json
{
  "author" : "Larry A. Gardner  extended author",
  "details" : "All function calls are currently implemented without side effects Repeated dev",
  "methods" : 
  {
    "age(uint256)" : 
    {
      "author" : "Mary A. Botanist",
      "details" : "The Alexandr N. Tetearing algorithm could increase precision",
      "params" : 
      {
        "rings" : "The number of rings from dendrochronological sample"
      },
      "return" : "age in years, rounded up for partial years"
    }
  },
  "title" : "A simulator for trees extended title"
}
```

and

```
{
  "methods" : 
  {
    "age(uint256)" : 
    {
      "notice" : "Calculate tree age in years, rounded up, for live trees"
    }
  },
  "notice" : "You can use this contract for only the most basic simulation         extended notice"
}
```

Reviewer: I think this was a misunderstanding. I was talking about the rst rendering, not about natspec.

facing messages. These messages may be shown to the end user (the human) at the
time that they will interact with the contract (i.e. sign a transaction).

It is recommended that Solidity contracts are fully annontated using NatSpec for
all public interfaces (everything in the ABI).

NatSpec includes the formatting for comments that the smart contract author will
use, and which are understood by the Solidity compiler. Also detailed below is

> output of the Solidity compiler, which extracts these comments into a machine-

Reviewer: Same here.

readable format.

.. _header-doc-example:

Documentation Example
=====================

Documentation is inserted above each ``class``, ``interface`` and
``function`` using the doxygen notation format.

-  For Solidity you may choose ``///`` for single or multi-line
   comments, or ``/**`` and ending with ``*/``.

-  For Vyper, use ``"""`` indented to the inner contents with bare
   comments. See `Vyper
   documentation <https://vyper.readthedocs.io/en/latest/structure-of-a-contract.html#natspec-metadata>`__.

The following example shows a contract and a function using all available tags.
Note: NatSpec currently does NOT apply to public state variables (see
`solidity#3418 <https://github.com/ethereum/solidity/issues/3418>`__),

> even if they are declared public and therefore do affect ABI. Note:

Reviewer: ```suggestion
even if they are declared public and therefore do affect the ABI. Note:
```

The Solidity compiler only interprets tags if they are external or
public. You are welcome to use similar comments for your internal and
private functions, but those will not be parsed.

.. code:: solidity

   pragma solidity ^0.5.6;

   /// @title A simulator for trees
   /// @author Larry A. Gardner
   /// @notice You can use this contract for only the most basic simulation
   /// @dev All function calls are currently implemented without side effects
   contract Tree {
       /// @author Mary A. Botanist
       /// @notice Calculate tree age in years, rounded up, for live trees
       /// @dev The Alexandr N. Tetearing algorithm could increase precision
       /// @param rings The number of rings from dendrochronological sample
       /// @return age in years, rounded up for partial years
       function age(uint256 rings) external pure returns (uint256) {
           return rings + 1;
       }
   }

.. _header-tags:

Tags
====

> All tags are optional. The following table explains the purpose of each

Reviewer: Do you know that there are people who re-formulate their sentences until all lines render with exactly the same width? It is the case for this paragraph, so I checked the others and was delighted to see that this just seems to be a coincidence :)

Author: Who can say that it was a coincidence!?

=========== =============================================================================== =============================
Tag                                                                                         Context
=========== =============================================================================== =============================
``@title``  A title that should describe the contract/interface                             contract, interface
``@author`` The name of the author                                                          contract, interface, function
``@notice`` Explain to an end user what this does                                           contract, interface, function
``@dev``    Explain to a developer any extra details                                        contract, interface, function
``@param``  Documents a parameter just like in doxygen (must be followed by parameter name) function
``@return`` Documents the return type of a contract's function                              function
=========== =============================================================================== =============================

> If your function returns a tuple, like ``(int quotient, int remainder)``

Reviewer: We don't really call these tuples in other parts of the documentation, so perhaps rather talk about "returns multiple variables like ``(int...`.

Author: Updated in 6667d17633ab6dc13a4e504e21292fa09163ea9c

then use multiple ``@return`` statements in the same format as the
``@param`` statements.

.. _header-dynamic:

Dynamic expressions
-------------------

> In function documentation, you may use dynamic expressions in any tags.

Reviewer: I think you should provide an example here, without context, "dynamic expressions" can mean anything. Perhaps also rather call them "evaluated expressions" or "expressions to be evaluated" or "templates"?

Author: Example and explanation provided in 9cfd68763bc8248140d2eba73b4e58bd40ae6082

The Solidity compiler will pass through these expressions and therefore
they are outside the scope of this documentation. You may read more at
`the radspec project <https://github.com/aragon/radspec>`__.

.. _header-inheritance:

Inheritance Notes
-----------------

Currently it is undefined whether a contract with a function having no
NatSpec will inherit the NatSpec of a parent contract/interface for that
same function.

.. _header-output:

Documentation Output
====================

When parsed by the compiler, documentation such as the one from the

> above example will produce 2 different JSON files. One is meant to be

Reviewer: ```suggestion
above example will produce two different JSON files. One is meant to be
```

Reviewer: Please change this, I think it hinders reading flow.

consumed by the end user as a notice when a function is executed and the
other to be used by the developer.

> If the above contract is saved as ex1.sol then you can generate the

Reviewer: ```suggestion
If the above contract is saved as ``ex1.sol`` then you can generate the
```

> Solidity contracts can have a special form of comments that form the

Reviewer: I think this file needs a heading.

> basis of the Ethereum Natural Specification Format.

Reviewer: This sounds like people already know what the Ethereum Natual Specification Format is. I think we should introduce the term here instead and explain what it is meant to be used for.

> .. _header-n3:

Reviewer: Please use a descriptive name.

Documentation Example
=====================

Documentation is inserted above each ``class``, ``interface`` and
``function`` using the doxygen notation format.

-  For Solidity you may choose ``///`` for single or multi-line
   comments, or ``/**`` and ending with ``*/``.

-  For Vyper, use ``"""`` indented to the inner contents with bare
   comments. See `Vyper
   documentation <https://vyper.readthedocs.io/en/latest/structure-of-a-contract.html#natspec-metadata>`__.

This example shows a contract and a function using all available tags.

> Note: NatSpec currently does NOT apply to variables (see

Reviewer: Are you talking about public state variables or variables in general?

`solidity#3418 <https://github.com/ethereum/solidity/issues/3418>`__),
even if they are declared public and therefore do affect ABI. Note:
NatSpec currently only interprets tags functions if they are external or
public. You are welcome to use similar comments for your internal and
private functions, but those will not be parsed.

.. code:: solidity

   pragma solidity ^0.5.1;

   /// @title A simulator for Bug Bunny, the most famous Rabbit
   /// @author Warned Bros
   /// @notice You can use this contract for only the most basic simulation
   /// @dev All function calls are currently implement without side effects

>    contract BugsBunny {

Reviewer: I'm not sure we should use trademarks (I assume) in the documentation without the need for them.

       /// @author Bob Clampett
       /// @notice Determine if Bugs will accept `(food)` to eat
       /// @dev String comparison may be inefficient
       /// @param food The name of a food to evaluate (English)
       /// @return true if Bugs will eat it, false otherwise
       function doesEat(string calldata food) external pure returns (bool) {
           return keccak256(bytes(food)) == keccak256("carrot");
       }
   }

> .. _header-n12:

Reviewer: Please use a descriptive name.

Tag                                                                                         Context
=========== =============================================================================== =============================
``@title``  A title that should describe the contract/interface                             contract, interface
``@author`` The name of the author                                                          contract, interface, function
``@notice`` Explain to an end user what this does                                           contract, interface, function
``@dev``    Explain to a developer any extra details                                        contract, interface, function
``@param``  Documents a parameter just like in doxygen (must be followed by parameter name) function
``@return`` Documents the return type of a contract's function                              function
=========== =============================================================================== =============================

If your function returns a tuple, like ``(int quotient, int remainder)``
then use multiple ``@return`` statements in the same format as the
``@param`` statements.

> .. _header-n44:

Reviewer: Please use a descriptive name.

Dynamic expressions
-------------------

> In function documentation, you may use dynamic expressions for all tags.

Reviewer: I think we should also mention the new effort of a more restrictive feature set in these expressions.

Example:

.. code::

>    /// @notice Send `(valueInmGAV / 1000).fixed(0,3)` GAV from the account of

Reviewer: It would be great if this could refer to the example above and not some weird coin nobody has heard of before ;)

Reviewer: In addition to the example, things like `.fixed(...)` have never worked.

   /// `message.caller.address()` to an account accessible only by `to.address()`
   function send(address to, uint256 valueInmGAV) {
   ...

If a user (address 0x2334) attempts to call this function with a ``to``

> address of ``0x0`` and ``valueInmGAV`` of 4,135 then this will render to

Reviewer: What is the meaning of the `,` in `4,134`? Any reason you use `,` here and `.` below?

the user as:

   Send 4.135 GAV from the account of 0.2334 to an account accessible
   only by 0x0

> Use any Javascript/Paperscript expression encapsulated in backticks as

Reviewer: I stil don't know what Paperscript was meant to refer to...

##############
NatSpec Format
##############

Solidity contracts can use a special form of comments to provide rich
documentation for functions, return variables and more. This special form is

> named the Ethereum Natural Specification Format (NatSpec).

Reviewer: ```suggestion
named the Ethereum Natural Language Specification Format (NatSpec).
```

Reviewer: I think we should already say here that it follows JavaDoc/DoxyGen style.

This documentation is segmented into developer-focused messages and end-user-
facing messages. These messages may be shown to the end user (the human) at the
time that they will interact with the contract (i.e. sign a transaction).

It is recommended that Solidity contracts are fully annontated using NatSpec for
all public interfaces (everything in the ABI).

NatSpec includes the formatting for comments that the smart contract author will
use, and which are understood by the Solidity compiler. Also detailed below is
output of the Solidity compiler, which extracts these comments into a machine-
readable format.

.. _header-doc-example:

Documentation Example
=====================

Documentation is inserted above each ``class``, ``interface`` and
``function`` using the doxygen notation format.

-  For Solidity you may choose ``///`` for single or multi-line
   comments, or ``/**`` and ending with ``*/``.

-  For Vyper, use ``"""`` indented to the inner contents with bare
   comments. See `Vyper
   documentation <https://vyper.readthedocs.io/en/latest/structure-of-a-contract.html#natspec-metadata>`__.

> This example shows a contract and a function using all available tags.

Reviewer: ```suggestion
The following example shows a contract and a function using all available tags.
```
(it is not that close to the sentence)

Note: NatSpec currently does NOT apply to public state variables (see
`solidity#3418 <https://github.com/ethereum/solidity/issues/3418>`__),
even if they are declared public and therefore do affect ABI. Note:

> NatSpec currently only interprets tags functions if they are external or

Reviewer: I think a word is missing here: "interprets tags functions"

Author: Updated in c476471

public. You are welcome to use similar comments for your internal and

> private functions, but those will not be parsed.

Reviewer: At this point, we do not yet know what it means for natspec comments to be parsed, do we?

Author: Updated in c476471

.. code:: solidity

   pragma solidity ^0.5.6;

   /// @title A simulator for trees
   /// @author Larry A. Gardner
   /// @notice You can use this contract for only the most basic simulation
   /// @dev All function calls are currently implemented without side effects
   contract Tree {
       /// @author Mary A. Botanist
       /// @notice Calculate tree age in years, rounded up, for live trees
       /// @dev The Alexandr N. Tetearing algorithm could increase precision
       /// @param rings The number of rings from dendrochronological sample
       /// @return age in years, rounded up for partial years
       function age(uint256 rings) external pure returns (uint256) {
           return rings + 1;
       }
   }

.. _header-tags:

Tags
====

All tags are optional. The following table explains the purpose of each

> NatSpec tag and where it may be used.

Reviewer: If no tag is given, `@notice` is assumed.

Author: Updated in 8233930

     "methods": {
       "send(address,uint256)": { "notice": "Send `(valueInmGAV / 1000).fixed(0,3)` GAV from the account of `message.caller.address()`, to an account accessible only by `to.address()`." },
       "balance(address)": { "notice": "`(balanceInmGAV / 1000).fixed(0,3)` GAV is the total funds available to `who.address()`." }
     },
     "invariants": [
       { "notice": "The sum total amount of GAV in the system is 1 million." }
     ],
     "construction": [
       { "notice": "Endows `message.caller.address()` with 1m GAV." }
     ]
   }

Note that the key by which to find the methods is the function's
canonical signature as defined in the `Contract
ABI <Ethereum-Contract-ABI#signature>`__ and not simply the function's
name.

.. _header-developer-doc:

Developer Documentation
-----------------------

Apart from the user documentation file, a developer documentation JSON
file should also be produced and should look like this:

.. code::

   {
     "author": "Larry A. Gardner",
     "title": "A simulator for trees",
     "methods": {
       "age(uint256)": {
         "details": "The Alexandr N. Tetearing algorithm could increase precision"
       }
     },

>      "invariants": [

Reviewer: Is such a concept present somewhere?

Author: Yes, this is a real thing. https://en.wikipedia.org/wiki/Dendrochronology

I went out to learn some biology rather than show some useless contrived example.

Reviewer: No, I mean the invariant concept - is that implemented in one of the languages that support natspec?

Reviewer: About the example: Yes, I really like it, too!

Author: Thanks! Ah, no the invariant is cruft. I need to rerun the program to get the correct output. Patch coming...

Author: Updated in 5c0eabe
